Maple2.Server.Game/Manager/ItemBoxManager.cs

```diff
@@ -0,0 +1,259 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Maple2.Database.Storage;
+using Maple2.Model;
+using Maple2.Model.Enum;
+using Maple2.Model.Error;
+using Maple2.Model.Game;
+using Maple2.Model.Metadata;
+using Maple2.Server.Game.Session;
+
+namespace Maple2.Server.Game.Manager;
+
+public class ItemBoxManager {
+    private readonly GameSession session;
+    public int BoxCount { get; private set; }
+
+    public ItemBoxManager(GameSession session) {
+        this.session = session;
+        BoxCount = 0;
+    }
+
+    public void Reset() {
+        BoxCount = 0;
+    }
+
+    public ItemBoxError Open(Item item, int count = 1, int index = 0) {
+        int[] itemBoxParams = item.Metadata.Function?.Parameters.Split(',').Select(int.Parse).ToArray() ?? Array.Empty<int>();
+        return item.Metadata.Function?.Type switch {
+            ItemFunction.SelectItemBox => SelectItemBox(item, itemBoxParams[0], itemBoxParams[1], index, count),
+            ItemFunction.OpenItemBox => OpenItemBox(item, itemBoxParams[0], itemBoxParams[1], itemBoxParams[2], itemBoxParams[3], itemBoxParams.Length == 5 ? itemBoxParams[4] : 1, count),
+            ItemFunction.OpenItemBoxWithKey => OpenItemBoxWithKey(item, itemBoxParams[0], itemBoxParams[1], itemBoxParams[2], itemBoxParams[5], count),
+            _ => throw new ArgumentOutOfRangeException(item.Metadata.Function?.Type.ToString(), "Invalid box type"),
+        };
+    }
+
+    private ItemBoxError SelectItemBox(Item item, int itemRequiredAmount, int boxId, int index, int count = 1) {
+        var error = ItemBoxError.ok;
+
+        // check if box count is enough to open
+        if (session.Item.Inventory.Find(item.Id).Sum(box => box.Amount) < itemRequiredAmount) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        // check if we can get metadata
+        if (!session.TableMetadata.IndividualItemDropTable.Entries.TryGetValue(boxId, out Dictionary<byte, IList<IndividualItemDropTable.Entry>>? dropGroupTable)) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        // SelectItemBox disregards DropGroup
+        IList<IndividualItemDropTable.Entry> drops = session.ItemBox.FilterDrops(dropGroupTable.Values.SelectMany(x => x).ToList());
+
+        if (drops.Count == 0) {
+            return ItemBoxError.s_err_cannot_open_multi_itembox_inventory_fail;
+        }
+
+        IndividualItemDropTable.Entry selectedEntry = drops[index];
```

Comment:
This seems unsafe, probably should use `ElementAtOrDefault` to handle out of bounds. Can just do this single null check instead of the above `drops.Count == 0` check.